contents/ops/ops.qmd

```diff
@@ -201,10 +205,37 @@ Teams build robust rollback procedures in case issues emerge. Rollbacks revert t
+For example, a retailer containerizes a product recommendation model in [TensorFlow Serving](https://www.tensorflow.org/tfx/guide/serving) and deploys it to a [Kubernetes](https://kubernetes.io/) staging cluster. After monitoring and approving performance on sample traffic, Kubernetes shifts 10% of production traffic to the new model. If no issues are detected after a few days, the new model takes over 100% of traffic. However, teams should keep the previous version accessible for rollback if needed.
 
 Model deployment processes enable teams to make ML systems resilient in production by accounting for all transition states.
 
+### Model Serving
+
+After model deployment, ML-as-a-Service becomes a critical component in the MLOps lifecycle. Online services such as Facebook/Meta handle tens of trillions of inference queries per day [@wu2019machine]. Model serving bridges the gap between developed models and ML applications or end-users, ensuring that deployed models are accessible, performant, and scalable in production environments.
+
+Several frameworks facilitate model serving, including [TensorFlow Serving](https://www.tensorflow.org/tfx/guide/serving), [NVIDIA Triton Inference Server](https://developer.nvidia.com/triton-inference-server), and [KServe](https://kserve.github.io/website/latest/) (formerly KFServing). These tools provide standardized interfaces for serving deployed models across various platforms and handle many complexities of model inference at scale.
+
+Model serving can be categorized into three main types:
+
+1. **Online Serving:** Provides real-time predictions with low latency, which is crucial for applications like recommendation systems or fraud detection.
+2. **Offline Serving:** Processes large batches of data asynchronously, suitable for tasks like periodic report generation.
+3. **Near-Online (semi-synchronous) Serving:** Balances between online and offline, offering relatively quick responses for less time-sensitive applications such as chatbots.
+
+One of the key challenges for model serving systems is operating under performance requirements defined by Service Level Agreements (SLAs) and Service Level Objectives (SLOs). SLAs are formal contracts specifying expected service levels. These service levels rely on metrics such as response time, availability, and throughput. SLOs are internal goals teams set to meet or exceed their SLAs. 
+
+For ML model serving, the SLA and SLO agreements and objectives directly impact user experience, system reliability, and business outcomes. Therefore, teams carefully tune their serving platform. ML serving systems employ various techniques to optimize performance and resource utilization, such as the following:
+
+I understand you'd like to combine the bullet points with the references to create a more concise list. Here's a revised version that incorporates the references into the bullet points:
```

Comment:
Remove? This line sounds like a comment

Reply:
Thanks for catching this. 

Just as FYI, originally I expanded out the paper details but then decided to collapse it because it was getting too long and the other sections weren't as balanced with this new level of detail. But just jotting down what I was working off of your draft just in case we want to resurrect it in the future: 

> Several works have contributed to these advancements in ML serving systems. @crankshaw2017clipper for instance introduce [Clipper](https://rise.cs.berkeley.edu/projects/clipper/), which is a low-latency online prediction serving system designed to meet the demands of real-time, accurate, and robust predictions under heavy query load. It introduces a modular architecture to simplify model deployment across frameworks and applications, and incorporates caching, batching, and adaptive model selection techniques to reduce prediction latency and improve throughput, accuracy, and robustness without modifying underlying machine learning frameworks. Clockwork [@gujarati2020serving] focuses on delivering high-performance model serving by predicting execution times of individual inferences and making efficient use of costly hardware accelerators.
> 
> @romero2021infaas further explored model instance selection and autoscaling in their [INFaaS framework](https://github.com/stanford-mast/INFaaS). INFaaS generates model-variants from already trained models and efficiently navigates the trade-off space of model-variants to select the appropriate model, hardware architecture, and optimizations for each query. Developers can simply specify the performance and accuracy requirements for their applications, and INFaaS handles the selection process. @zhang2019mark show how to do load balancing and autoscaling in their MArk (Model
> Ark) system. @li2023alpaserve demonstrated advanced model orchestration techniques in their AlpaServe framework, enabling efficient handling of large models and complex serving scenarios.